masonry_core/src/app/render_root.rs

```diff
@@ -287,6 +306,80 @@ pub(crate) struct InspectorState {
 }
 
 impl RenderRoot {
+    /// Compute a focus fallback according to the provided `policy`.
+    ///
+    /// Returns `Some(id)` if a suitable widget is found, otherwise `None`.
+    pub fn compute_focus_fallback(&self, policy: FocusFallbackPolicy) -> Option<WidgetId> {
+        let base = self.layer_root_id(0);
+        match &policy {
+            FocusFallbackPolicy::AnchorSubtreeFirst(inner) => {
+                if let Some(anchor) = self.global_state.focus_anchor
+                    && self.is_still_interactive(anchor)
+                    && let Some(id) = self.match_policy_at(inner, anchor)
+                {
+                    return Some(id);
+                }
+                self.match_policy_at(inner, base)
+            }
+            FocusFallbackPolicy::FirstTextInput
+            | FocusFallbackPolicy::FirstFocusable
+            | FocusFallbackPolicy::Tagged(_) => self.match_policy_at(&policy, base),
+        }
+    }
+
+    /// Depth-first pre-order search from `start` for first widget whose `predicate` returns true.
+    fn find_first_matching(
+        &self,
+        start: WidgetId,
+        predicate: impl Fn(&WidgetRef<'_, dyn Widget>) -> bool,
+    ) -> Option<WidgetId> {
+        let mut stack = vec![self.get_widget(start)?];
```

Comment:
I guess the reason we've not done a recursive search here is because the function can't be passed to the children? In other similar functions (e.g. `find_widget_by_id`), we've done a recursive search rather than allocating.

I believe that you can either accept the predicate by reference, or ensure that the predicate is `Copy` (which incidentally is equivalent but more ergonomic, as I believe that a non-copy predicate can just be borrowed to make it Copy)